### ISSUE TITLE
cannot login anymore

### ISSUE THREAD

**Issue by derStephan** [nuxsmin/sysPass#536]
Hi,

I tried to login into my syspass 2.1 yesterday. This failed with the message 

"No es posible acceder directamente a este archivo"

Normally, after entering a password, it is converted in something quite long. But now it is converted into 5 characters only. 

I used "webdeveloper" to show the password field and it says "false". 

3 Days ago, I could login without any problem. I did not change anything.

**Comment by nuxsmin** [comment]
Hi, where is that message shown?

**Comment by nuxsmin** [comment]
That message is shown when you try to access to a PHP file without initialization, that is pointing to "somefile.php" directly.

**Comment by nuxsmin** [comment]
Please, reload the browser  page (CTRL+F5)

**Comment by derStephan** [comment]
I did not try to access another file directly. I just open the index.php as always.

I reloaded the browser and renewed the cache. I even deleted all cookies with no success. 

When I access it from the Internet Explorer 11, this message does not show up but after login, I get a blank page.

**Comment by nuxsmin** [comment]
So the index page is shown fine, so it's when logging in, right?

After analyzing some code, the message is most likely shown when sending a form, because every request should be GET or POST and the code is checking if the form is using the right type.

Some versions ago the Javascript code was changed and some requests were changed to a different type (GET or POST), so either the Javascript or app code aren't updated, the check will fail.

To sum up, if the app expects "GET", Javascript should request through "GET", and the same for "POST"

**Comment by nuxsmin** [comment]
The code that checks the request is:

```
    public static function checkReferer($method)
    {
        $referer = self::getRequestHeaders('HTTP_REFERER');

        if (!$referer
            || $_SERVER['REQUEST_METHOD'] !== strtoupper($method)
            || !preg_match('#' . Init::$WEBROOT . '/.*$#', $referer)
        ) {
            Init::initError(__('No es posible acceder directamente a este archivo'));
            exit();
        }
    }
``

If the `HTTP_REFERER` header is not sent the check will fail too. Are you using a proxy?

**Comment by derStephan** [comment]
>If the `HTTP_REFERER` header is not sent the check will fail too. Are you using a proxy?

Ah! That was the problem. I recently deactivated referrer due to privacy reasons. After switching it on again, I can login again. 

Why do you check the referrer? Maybe, the error could be clearer in these cases?

**Comment by nuxsmin** [comment]
Great!

The HTTP_REFERER is checked because every Ajax request should have it for checking if the request is sent from the original URL:

`!preg_match('#' . Init::$WEBROOT . '/.*$#', $referer)`

**Comment by nuxsmin** [comment]
Usually, this kind of validation takes place in the web browser (CORS is denied by default for Ajax requests), but you can use some tools like wget, curl, etc that can send direct requests to those scripts.
